src/Agent.Worker/Build/BuildJobExtension.cs

```diff
@@ -38,24 +49,93 @@ private Task PrepareAsync()
             IExecutionContext executionContext = PrepareStep.ExecutionContext;
             var directoryManager = HostContext.GetService<IBuildDirectoryManager>();
 
+            // This flag can be false for jobs like cleanup artifacts.
+            // If syncSources = false, we will not set source related build variable, not create build folder, not sync source.
+            bool syncSources = executionContext.Variables.Build_SyncSources ?? true;
+            if (!syncSources)
+            {
+                Trace.Verbose($"{Constants.Variables.Build.SyncSources} = false, we will not set source related build variable, not create build folder and not sync source");
+                executionContext.Output($"{Constants.Variables.Build.SyncSources} = false, skip get source");
+                return;
+            }
+
             // Get the repo endpoint and source provider.
-            ServiceEndpoint endpoint = null;
-            ISourceProvider sourceProvider = null;
-            if (!TryGetPrimaryEndpointAndSourceProvider(executionContext, out endpoint, out sourceProvider))
+            if (!TryGetPrimaryEndpointAndSourceProvider(executionContext, out _sourceEndpoint, out _sourceProvider))
             {
                 throw new Exception(StringUtil.Loc("SupportedRepositoryEndpointNotFound"));
             }
 
+            executionContext.Debug($"Primary repository: {_sourceEndpoint.Name}. repository type: {_sourceProvider.RepositoryType}");
+
             // Prepare the build directory.
+            executionContext.Debug("Preparing build directory.");
             TrackingConfig trackingConfig = directoryManager.PrepareDirectory(
                 executionContext,
-                endpoint,
-                sourceProvider);
+                _sourceEndpoint,
+                _sourceProvider);
+
+            executionContext.Debug("Set build variables.");
+            string _workDirectory = IOUtil.GetWorkPath(HostContext);
+            executionContext.Variables.Set(Constants.Variables.System.DefaultWorkingDirectory, Path.Combine(_workDirectory, trackingConfig.SourcesDirectory));
+            executionContext.Variables.Set(Constants.Variables.Common.TestResultsDirectory, Path.Combine(_workDirectory, trackingConfig.TestResultsDirectory));
+
+            executionContext.Variables.Set(Constants.Variables.Build.BinariesFolder, Path.Combine(_workDirectory, trackingConfig.BuildDirectory, Constants.Build.Path.BinariesDirectory));
+            executionContext.Variables.Set(Constants.Variables.Build.BuildFolder, Path.Combine(_workDirectory, trackingConfig.BuildDirectory));
+            executionContext.Variables.Set(Constants.Variables.Build.ArtifactsDirectory, Path.Combine(_workDirectory, trackingConfig.BuildDirectory));
```

Comment:
Or is this supposed to be SYSTEM_ARTIFACTSDIRECTORY? Whatever the correct thing here, might be a worth adding a comment if it's confusing.